### DISCUSSION TITLE
Apply Mermaid config settings fails

### DISCUSSION THREAD

**Discussion by aghaynes** [quarto-dev/quarto-cli#4250] [Q&A] [answered]
Hi!

It's possible to modify e.g. distance between nodes in Mermaid flowcharts via a config at the beginning of the graph definition (since v8.6.0), e.g. 

```
graph TD
%%{init:{'flowchart':{'nodeSpacing': 20, 'rankSpacing': 30}}}%%
A[Christmas] -->|Get money| B(Go shopping)
B --> C{Let me think}
C -->|One| D[Laptop]
C -->|Two| E[iPhone]
C -->|Three| F[fa:fa-car Car]
```

https://github.com/mermaid-js/mermaid/issues/1312

On the live editor:
https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNpFkMFqwzAMhl9FGEI2aKCwy8hhsCZdL4UN2lvcg0iU2jSWjeNQSuJ3nzu29SZ9-n6BNIvWdiRKcfboFBxryVk2a9ahnPN-sNdWoQ95ajh5B4et5nNewut6BblHvjzQyzrGmGWS35tKeT0Gg-MJiuJt2VEAY5luC2yedhZGZZ1LoWfJm7sA1by_KwRBab5EydVP7pNpgbrZowvWnf7o8WoX2Db6S6WVD6o8Jfuj6bHssWjRQ4U-jcVKGPIGdZeunCUDSBEUGZKiTGVHPU5DkEJyTCpOwR5u3Ioy-IlWYnIdBqo1pv8YkTYP4z_ddjpY_wvjN6uqb-Y

This just produces syntax error in Quarto... I guess the %% is causing the line to be treated as a comment...?
Is it possible to modify e.g. nodeSpacing some other way in Quarto?

Thanks!

**Comment by cscheid** [comment, marked as answer]
I can't reproduce on the latest version of quarto:

<img width="930" alt="image" src="https://user-images.githubusercontent.com/285675/217667695-be6eea1a-497e-4834-9d9b-ab95393fdf06.png">

**Reply by aghaynes** [reply]
OK, thanks for looking... I'm using the version bundled with RStudio 2022.12.0... maybe it's just that then...

On a related note, is it possible to use R to put values into the nodes?  e.g. D[Laptop (n = `r nrow(laptops)`)]

**Reply by cscheid** [reply]
No, it's not (at least not without having to write a large amount of custom javascript). Mermaid diagrams are rendered before knitr invocations.

**Reply by aghaynes** [reply]
OK, thanks... would have to use this then I suppose... https://rich-iannone.github.io/DiagrammeR/mermaid.html

**Reply by harrelfe** [reply]
Here is an approach to inserting R calculations into `mermaid` charts: https://hbiostat.org/rflow/doverview.html#sec-doverview-filter

**Reply by aghaynes** [reply]
Thanks Frank, but you're getting syntax errors too... ;)
![grafik](https://user-images.githubusercontent.com/6594028/218431024-2c2ea0e9-104e-4e99-b27a-9f83326691b0.png)

**Reply by LJFRIESE** [reply]
For anyone in the future searching for embedding r objects into mermaid diagrams:

You can use `makemermaid` from @harrelfe's package: https://hbiostat.org/r/qreport/. This is probably pretty robust, I've not used it.

Or, you can read the mermaid syntax into R as a string, use interpolation (like `glue::glue()`), then knit. There's more than one (better) way, but here's an example. The example also includes how to use the config syntax described in [the mermaid docs](https://mermaid.js.org/config/theming.html#customizing-themes-with-themevariables). The quarto docs touch on this, but don't have much detail.

## File 1 (Parent):

```
---
title: "mermaid_example"
format: html
---
```{r, results='asis', echo=FALSE}
# results='asis' is required to render the child in the parent

# Variables to include
variable1 <- "1000"
variable2 <- "Look ma, interpolated vars in a mermaid!"
src <- knitr::knit_expand(text = glue::glue(paste0(readLines("test__.qmd"),collapse="\n"),
                                            .open="<",
                                            .close=">",
                                            # Pass the variables to glue's environment
                                            .envir=rlang::env(
                                              variable1 = variable1,
                                              variable2 = variable2)
                                            ))
# Render the mermaid diagram
  knitr::knit_child(text = src,
                     options = list(echo=FALSE,
                                    results='asis'),
                     quiet = TRUE,
   ) |> # cat to the parent md
     cat()
`` # Three backticks here... markdown isn't letting me escape them in the code block
```

## File 2 (child)


```
```{mermaid}
%%| fig-height: 5
%%{
init: {
  "flowchart": {"nodeSpacing": 200, "rankSpacing": 300, "useMaxWidth": "true", "htmlLabels": "true", "curve": 'cardinal'},
  "securityLevel": "loose",
  "themeVariables": { 'primaryColor': '#BB2528',
                      'primaryTextColor': '#fff',
                      'primaryBorderColor': '#7C0000',
                      'lineColor': '#F8B229',
                      'secondaryColor': '#006100',
                      'tertiaryColor': '#fff'},
  "theme": "base",
    "themeCSS": ".extension { fill: white !important; };.divider { display: none; };"
    
}}%%
flowchart
  A[Christmas] -->|<variable1>| B(Go shopping)
  B --> C{<variable2>}
  B --> G[/Another/]
  C ==>|One| D[Laptop]
  C -->|Two| E[iPhone]
  C -->|Three| F[fa:fa-car Car]
  subgraph section
    C
    D
    E
    F
    G
  end

`` # Three backticks here... markdown isn't letting me escape them in the code block
```


Output:

![image](https://github.com/quarto-dev/quarto-cli/assets/135739244/17de8f98-c082-4f76-b623-078b15ec1182)

**Reply by mcanouil** [reply]
In markdown to escape backticks you only need to use more.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````
